src/erasure.rs

```diff
@@ -161,263 +168,418 @@ pub fn decode_blocks(data: &mut [&mut [u8]], coding: &[&[u8]], erasures: &[i32])
-                        data_ptrs.len()
-                    );
-                    decode_blocks(data_ptrs.as_mut_slice(), &coding_ptrs, &erasures)?;
-                }
-                for i in &erasures[..erasures.len() - 1] {
-                    let idx = *i as usize;
-                    let data_size = locks[idx].get_data_size().unwrap() - BLOB_HEADER_SIZE as u64;
-                    locks[idx].meta = meta.clone().unwrap();
-                    locks[idx].set_size(data_size as usize);
-                    trace!(
-                        "erasures[{}] size: {} data[0]: {}",
-                        *i,
-                        data_size,
-                        locks[idx].data()[0]
-                    );
-                }
+    }
+
+    // marks end of erasures
+    erasures.push(-1);
+    trace!(
+        "erasures[]: {:x} {:?} data_size: {}",
+        debug_id,
+        erasures,
+        size.unwrap(),
+    );
+    //lock everything for write
+    for b in &blobs {
+        locks.push(b.write().expect("'locks' arr in pb fn recover"));
+    }
+
+    {
+        let mut coding_ptrs: Vec<&mut [u8]> = Vec::with_capacity(NUM_CODING);
+        let mut data_ptrs: Vec<&mut [u8]> = Vec::with_capacity(NUM_DATA);
+        for (i, l) in locks.iter_mut().enumerate() {
+            if i < NUM_DATA {
+                trace!("{:x} pushing data: {}", debug_id, i);
+                data_ptrs.push(&mut l.data[..size.unwrap()]);
+            } else {
+                trace!("{:x} pushing coding: {}", debug_id, i);
+                coding_ptrs.push(&mut l.data_mut()[..size.unwrap()]);
             }
         }
-        block_start += NUM_CODED;
+        trace!(
+            "{:x} coding_ptrs.len: {} data_ptrs.len {}",
+            debug_id,
+            coding_ptrs.len(),
+            data_ptrs.len()
+        );
+        decode_blocks(
+            data_ptrs.as_mut_slice(),
+            coding_ptrs.as_mut_slice(),
+            &erasures,
+        )?;
+    }
+
+    let mut corrupt = false;
+    // repopulate header data size from recovered blob contents
+    for i in &erasures[..erasures.len() - 1] {
+        let n = *i as usize;
+        let mut idx = n as u64 + block_start_idx;
+
+        let mut data_size;
+        if n < NUM_DATA {
+            data_size = locks[n].get_data_size().unwrap();
+            data_size -= BLOB_HEADER_SIZE as u64;
+        } else {
+            data_size = size.unwrap() as u64;
+            idx -= NUM_CODING as u64;
+            locks[n].set_index(idx).unwrap();
+        }
+
+        locks[n].meta = meta.clone().unwrap();
+        locks[n].set_size(data_size as usize);
+        trace!(
+            "{:x} erasures[{}] ({}) size: {:x} data[0]: {}",
+            debug_id,
+            *i,
+            idx,
+            data_size,
+            locks[n].data()[0]
+        );
+        if data_size > BLOB_SIZE as u64 {
+            corrupt = true;
+        }
     }
+    assert!(!corrupt, " {:x} ", debug_id);
```

Comment:
Why assert here and not return an error?

Reply:
the error is currently unrecoverable, we've already polluted the window with poison blobs...

covered in #742